app/domain/invoice/batch_create.rb

```diff
@@ -76,10 +80,38 @@ def update_invoice_list
 
   def attributes(recipient)
     invoice.attributes.merge(
-      invoice_items_attributes: invoice.invoice_items.collect(&:attributes),
+      invoice_items_attributes: invoice_items_attributes(recipient),
       recipient_id: recipient.id,
       invoice_list_id: invoice_list.id,
       creator_id: invoice_list.creator_id
     )
   end
+
+  def invoice_items_attributes(recipient)
+    invoice.invoice_items.collect do |item|
+      begin
+        if item.variable_donation?
+          item.unit_cost = variable_donation_amount(recipient)
+
+          raise NoDonationsPresentError.new(invoice_item_id: item.id) if item.unit_cost.zero?
+        end
+
+        item.attributes
+      rescue NoDonationsPresentError
+        raise NoDonationsPresentError if invoice.invoice_items.size == 1
+
+        next
+      end
```

Comment:
An sich sollte das `rescue` ohne `begin` und `end` in allen semantischen Blöcken möglich sein. Es kann aber sein, dass das erst nach Ruby 2.5 möglich ist. Probier es mal aus :-)